src/libwinpty/winpty.cc

```diff
@@ -494,24 +495,92 @@ static OwnedHandle startAgentProcess(
         sui.dwFlags |= STARTF_USESHOWWINDOW;
         sui.wShowWindow = SW_HIDE;
     }
+
     PROCESS_INFORMATION pi = {};
-    const BOOL success =
-        CreateProcessW(exePath.c_str(),
-                       cmdlineV.data(),
-                       nullptr, nullptr,
-                       /*bInheritHandles=*/FALSE,
-                       /*dwCreationFlags=*/creationFlags,
-                       nullptr, nullptr,
-                       &sui, &pi);
-    if (!success) {
-        const DWORD lastError = GetLastError();
-        const auto errStr =
-            (WStringBuilder(256)
-                << L"winpty-agent CreateProcess failed: cmdline='" << cmdline
-                << L"' err=0x" << whexOfInt(lastError)).str_moved();
-        throw LibWinptyException(
-            WINPTY_ERROR_AGENT_CREATION_FAILED, errStr.c_str());
+    if (cfg->flags & WINPTY_FLAG_IMPERSONATE_THREAD) {
+        HRESULT hr;
+        HANDLE token = nullptr;
+        if (!OpenThreadToken(
+            GetCurrentThread(),
+            TOKEN_QUERY | TOKEN_DUPLICATE | TOKEN_ASSIGN_PRIMARY,
+            FALSE,
+            &token))
+        {
+            const DWORD lastError = GetLastError();
+            const auto errStr =
+                (WStringBuilder(256)
+                    << L"winpty-agent OpenThreadToken failed: cmdline='" << cmdline
+                    << L"' err=0x" << whexOfInt(lastError)).str_moved();
+            throw LibWinptyException(
+                WINPTY_ERROR_AGENT_CREATION_FAILED, errStr.c_str());
+        }
+
+        HANDLE dupToken = nullptr;
+        if (!DuplicateTokenEx(
+            token,
+            TOKEN_QUERY | TOKEN_DUPLICATE | TOKEN_ASSIGN_PRIMARY,
+            NULL,
+            SecurityImpersonation,
+            TokenPrimary,
+            &dupToken))
+        {
+            CloseHandle(token);
+            const DWORD lastError = GetLastError();
+            const auto errStr =
+                (WStringBuilder(256)
+                    << L"winpty-agent DuplicateTokenEx failed: cmdline='" << cmdline
+                    << L"' err=0x" << whexOfInt(lastError)).str_moved();
+            throw LibWinptyException(
+                WINPTY_ERROR_AGENT_CREATION_FAILED, errStr.c_str());
+        }
+
+        DWORD exitCode = 0;
+        const BOOL success = CreateProcessAsUser(
+            dupToken,
+            exePath.c_str(),
+            cmdlineV.data(),
+            nullptr,
+            nullptr,
+            FALSE,
+            CREATE_DEFAULT_ERROR_MODE | CREATE_BREAKAWAY_FROM_JOB | CREATE_UNICODE_ENVIRONMENT | NORMAL_PRIORITY_CLASS,
```

Comment:
I'm curious about these creation flags.

 * startAgentProcess is already called with a creation flags parameter, which is currently only set to either `DETACHED_PROCESS` (for `--create-desktop`) or `CREATE_NEW_CONSOLE` (normal code path). This code doesn't ever use `CREATE_NEW_CONSOLE`, so I assume it reuses its parent's console, if it has one?
 * Regarding the specific flags:
   * `CREATE_DEFAULT_ERROR_MODE`: I wonder if this ought to be the default behavior for winpty -- there was a complaint a while ago about crash handling and winpty (https://github.com/Microsoft/vscode/issues/32134), but I never looked into it. Maybe this creation flag fixes the problem?
   * `CREATE_UNICODE_ENVIRONMENT`: `lpEnvironment` is null; does this do anything?
   * `CREATE_BREAKAWAY_FROM_JOB` and `NORMAL_PRIORITY_CLASS`: I wonder if these flags should really be set by `WINPTY_FLAG_IMPERSONATE_THREAD` or whether they're orthogonal settings. Maybe winpty needs a way to specify arbitrary creation flags for the agent and/or the agent's child?